pmlc/rt/runtime_registry.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020, Intel Corporation
+
+#include "pmlc/rt/runtime_registry.h"
+
+#include <list>
+#include <mutex>
+#include <stdexcept>
+#include <utility>
+
+#include "pmlc/util/logging.h"
+#include "llvm/Support/FormatVariadic.h"
+
+namespace pmlc::rt {
+namespace {
+
+struct LoaderList {
+  std::mutex mutex;
+  std::list<std::pair<std::string, Loader>> loaders;
+
+  static LoaderList &instance() {
+    static LoaderList ll;
+    return ll;
+  }
+};
+
+class MemoizedFactory final {
+public:
+  explicit MemoizedFactory(Factory factory) : factory{std::move(factory)} {}
+  Runtime *getRuntime() {
+    std::call_once(init_once, [this]() { runtime = factory(); });
+    return runtime.get();
+  }
+
+private:
+  std::once_flag init_once;
+  Factory factory;
+  std::shared_ptr<Runtime> runtime;
+};
+
+const std::unordered_map<std::string, std::function<Runtime *()>>
+buildRuntimeMap() {
```

Comment:
Can we just build this during `plaidml::exec::init()`?

Reply:
Definitely.